File: web/backend/src/main/java/com/reportengine/web/controller/ExportController.java
package com.reportengine.web.controller;

import com.reportengine.lib.exporter.ExcelExporter;
import com.reportengine.lib.exporter.PdfExporter;
import com.reportengine.lib.model.RenderRequest;
import io.swagger.v3.oas.annotations.Operation;
import io.swagger.v3.oas.annotations.tags.Tag;
import org.springframework.http.HttpHeaders;
import org.springframework.http.MediaType;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

/**
 * 导出控制器 - 提供 PDF/Excel 导出 API
 */
@RestController
@RequestMapping("/api/export")
@CrossOrigin(origins = {"http://localhost:5173", "http://localhost:3000", "http://127.0.0.1:3000"})
@Tag(name = "报表导出", description = "将模板渲染为 PDF 或 Excel 二进制流")
public class ExportController {

    private final PdfExporter pdfExporter = new PdfExporter();
    private final ExcelExporter excelExporter = new ExcelExporter();

    /**
     * 导出 PDF
     */
    @PostMapping(value = "/pdf", produces = MediaType.APPLICATION_PDF_VALUE)
    @Operation(
        summary = "导出 PDF",
        description = "将报表模板渲染为 PDF 字节流。模板内 {{page}} 表示当前页号，{{totalPages}} 表示总页数。"
    )
    public ResponseEntity<byte[]> exportPdf(@RequestBody RenderRequest request) {
        try {
            byte[] pdfBytes = pdfExporter.export(request);

            HttpHeaders headers = new HttpHeaders();
            headers.setContentType(MediaType.APPLICATION_PDF);
            headers.setContentDispositionFormData("attachment", "report.pdf");

            return ResponseEntity.ok().headers(headers).body(pdfBytes);
        } catch (Exception e) {
            e.printStackTrace();
            return ResponseEntity.badRequest().body(("Error: " + e.getMessage()).getBytes());
        }
    }

    /**
     * 导出 Excel
     */
    @PostMapping(value = "/excel",
        produces = "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet")
    @Operation(
        summary = "导出 Excel",
        description = "将报表模板渲染为 xlsx 字节流。支持精确布局、打印设置、页脚页码、字体配置。"
    )
    public ResponseEntity<byte[]> exportExcel(@RequestBody RenderRequest request) {
        try {
            byte[] excelBytes = excelExporter.export(request);

            HttpHeaders headers = new HttpHeaders();
            headers.setContentType(MediaType.parseMediaType("application/vnd.openxmlformats-officedocument.spreadsheetml.sheet"));
            headers.setContentDispositionFormData("attachment", "report.xlsx");

            return ResponseEntity.ok().headers(headers).body(excelBytes);
        } catch (Exception e) {
            e.printStackTrace();
            return ResponseEntity.badRequest().body(("Error: " + e.getMessage()).getBytes());
        }
    }
}

File: web/ReportEngine.WebApi/Controllers/ExportController.cs
using Microsoft.AspNetCore.Mvc;
using ReportEngine.Core.Parsing;
using ReportEngine.Core.Rendering;
using ReportEngine.Core.SubReports;
using ReportEngine.Export.Pdf;
using ReportEngine.Export.Excel;

namespace ReportEngine.WebApi.Controllers;

[ApiController]
[Route("api/[controller]")]
public class ExportController : ControllerBase
{
    private readonly TemplateParser _parser = new();
    private readonly InMemoryTemplateResolver _resolver = new();
    private readonly PdfSharpExporter _pdfExporter = new();
    private readonly ClosedXmlExporter _excelExporter = new();

    /// <summary>
    /// 导出 PDF
    /// </summary>
    [HttpPost("pdf")]
    public async Task<IActionResult> ExportPdf([FromBody] Models.RenderRequest request)
    {
        try
        {
            var template = _parser.Parse(request.TemplateJson);
            var dataSources = ConvertData(request.Data);

            var renderer = new ReportRenderer(_resolver);
            var rendered = await renderer.RenderAsync(template, dataSources);

            var pdfBytes = _pdfExporter.Export(rendered);

            return File(pdfBytes, "application/pdf", "report.pdf");
        }
        catch (Exception ex)
        {
            return BadRequest(new { error = ex.Message });
        }
    }

    /// <summary>
    /// 导出 Excel
    /// </summary>
    [HttpPost("excel")]
    public async Task<IActionResult> ExportExcel([FromBody] Models.RenderRequest request)
    {
        try
        {
            var template = _parser.Parse(request.TemplateJson);
            var dataSources = ConvertData(request.Data);

            var renderer = new ReportRenderer(_resolver);
            var rendered = await renderer.RenderAsync(template, dataSources);

            var excelBytes = _excelExporter.Export(rendered);

            return File(excelBytes, "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet", "report.xlsx");
        }
        catch (Exception ex)
        {
            return BadRequest(new { error = ex.Message });
        }
    }

    private Dictionary<string, List<Dictionary<string, object>>> ConvertData(
        Dictionary<string, List<Dictionary<string, object?>>> data)
    {
        var result = new Dictionary<string, List<Dictionary<string, object>>>();
        foreach (var kvp in data)
        {
            var rows = kvp.Value.Select(row =>
                row.ToDictionary(
                    k => k.Key,
                    v => v.Value ?? (object)""
                )).ToList();
            result[kvp.Key] = rows;
        }
        return result;
    }
}
